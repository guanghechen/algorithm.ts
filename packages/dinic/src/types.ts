import type { IDigraphEdge } from '@algorithm.ts/types'

export interface IDinicEdge extends IDigraphEdge {
  /**
   * The starting node of the arc.
   */
  from: number
  /**
   * The ending node of the arc.
   */
  to: number
  /**
   * Capacity of the arc.
   */
  cap: number
  /**
   * Flow on the arc.
   */
  flow: number
}

export interface IDinic {
  /**
   * Initialize the Dinic algorithm.
   * @param source  the source node
   * @param sink    the sink node
   * @param n       the number of nodes
   */
  init(source: number, sink: number, n: number): void
  /**
   * Add an edge into the residual network.
   * @param from    the starting node
   * @param to      the ending node
   * @param cap     capacity
   */
  addEdge(from: number, to: number, cap: number): void
  /**
   * Calculate the maximum flow of the residual network.
   */
  maxflow(): number
  /**
   * Calculate the minium cut of the residual network.
   */
  mincut(): Array<Readonly<IDinicEdge>>
}
