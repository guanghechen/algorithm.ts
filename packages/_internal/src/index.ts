export * from './constant'
export * from './types'
