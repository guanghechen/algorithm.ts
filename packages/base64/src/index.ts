import { Base64 } from './base64'

export * from './base64'

const base64 = new Base64()

export const encode = base64.encode.bind(base64)
export const decode = base64.decode.bind(base64)
export const validate = base64.validate.bind(base64)
export default { encode, decode, validate }
