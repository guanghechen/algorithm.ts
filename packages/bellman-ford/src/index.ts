import { createCircularQueue } from '@algorithm.ts/circular-queue'

export interface IEdge {
  /**
   * The other end of the edge.
   */
  to: number
  /**
   * The cost of walking along this side.
   */
  cost: number
}

export interface IGraph {
  /**
   * The number of nodes in the graph. (0-index)
   */
  N: number
  /**
   * The source node. (0-index)
   */
  source: number
  /**
   * Graph edges.
   */
  edges: ReadonlyArray<IEdge>
  /**
   * Adjacency list. G[i] represent the index list of the  edges start from node i.
   */
  G: ReadonlyArray<ReadonlyArray<number>>
  /**
   * An array recording the shortest distance to the source point.
   */
  dist: number[]
}

export interface IOptions {
  /**
   * A big number, representing the unreachable cost.
   */
  INF?: number
  /**
   * Used to check if an element is already in the queue.
   */
  inq?: boolean[]
  /**
   * Record the number of times an element is enqueued,
   * used to check whether there is a negative cycle.
   */
  inqTimes?: number[]
}

const ZERO = 0
const DEFAULT_INF = Math.floor(Number.MAX_SAFE_INTEGER / 2)
const Q = createCircularQueue<number>()

/**
 * The bellman-ford algorithm, optimized with queue.
 *
 * @param graph
 * @param options
 */
export function bellmanFord(graph: IGraph, options: IOptions = {}): boolean {
  const { N, source, edges, G, dist } = graph
  const { INF = DEFAULT_INF, inq = [], inqTimes = [] } = options

  if (dist.length < N) dist.length = N
  if (inq.length < N) inq.length = N
  if (inqTimes.length < N) inqTimes.length = N

  dist.fill(INF, 0, N)
  inq.fill(false, 0, N)
  inqTimes.fill(0, 0, N)

  Q.init(N + 1)
  Q.enqueue(source)
  dist[source] = ZERO

  while (Q.size() > 0) {
    const o: number = Q.dequeue()!
    inq[o] = false

    for (const idx of G[o]) {
      const edge: IEdge = edges[idx]
      if (dist[edge.to] > dist[o] + edge.cost) {
        dist[edge.to] = dist[o] + edge.cost
        if (!inq[edge.to]) {
          Q.enqueue(edge.to)
          inq[edge.to] = true

          // eslint-disable-next-line no-plusplus
          if (++inqTimes[edge.to] > N) return false
        }
      }
    }
  }
  return true
}

export default bellmanFord
