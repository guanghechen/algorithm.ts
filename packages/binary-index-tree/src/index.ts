export * from './tree1'
export * from './tree2'
export * from './types'
export * from './util'
