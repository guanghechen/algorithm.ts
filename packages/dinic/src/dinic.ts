import type { ICircularQueue } from '@algorithm.ts/queue'
import { CircularQueue } from '@algorithm.ts/queue'
import type { IDinic, IDinicEdge } from './types'

export class Dinic implements IDinic {
  protected readonly _cur: number[] = [] // The next edge number to be considered of the edges starting from the i-th node.
  protected readonly _dist: number[] = [] // The distance from the source node to the i-th node.
  protected readonly _edges: IDinicEdge[] = []
  protected readonly _G: number[][] = []
  protected readonly _Q: ICircularQueue<number> = new CircularQueue()
  protected _N: number // The number of nodes in a network flow
  protected _source: number // The source point in a network flow
  protected _sink: number // The sink in a network flow
  protected _maxflow: number
  protected _edgesTot: number
  protected _modifiedTimestamp: number
  protected _resolvedTimestamp: number

  constructor() {
    this._N = 0
    this._source = -1
    this._sink = -1
    this._maxflow = 0
    this._edgesTot = 0
    this._modifiedTimestamp = 0
    this._resolvedTimestamp = -1
  }

  public init(source: number, sink: number, n: number): void {
    this._N = n
    this._source = source
    this._sink = sink
    this._maxflow = 0
    this._edgesTot = 0
    this._modifiedTimestamp = 0
    this._resolvedTimestamp = -1

    // Resize arrays.
    this._cur.length = n
    this._dist.length = n
    this._G.length = n

    this._Q.init()
    this._Q.resize(n + 1)
    for (let i = 0; i < n; i++) this._G[i] = []
  }

  public addEdge(from: number, to: number, cap: number): void {
    const { _G, _edges, _edgesTot } = this
    _G[from].push(_edgesTot)
    _G[to].push(_edgesTot + 1)
    _edges[_edgesTot] = { from, to, cap, flow: 0 }
    _edges[_edgesTot + 1] = { from: to, to: from, cap: 0, flow: 0 }

    this._edgesTot += 2
    this._modifiedTimestamp += 1
  }

  public maxflow(): number {
    if (this._resolvedTimestamp < this._modifiedTimestamp) {
      let { _maxflow } = this
      const { _N, _source, _cur } = this
      while (this._bfs()) {
        _cur.fill(0, 0, _N)
        _maxflow += this._dfs(_source, Number.MAX_SAFE_INTEGER)
      }

      this._maxflow = _maxflow
      this._resolvedTimestamp = this._modifiedTimestamp
    }
    return this._maxflow
  }

  public mincut(): Array<Readonly<IDinicEdge>> {
    this.maxflow()
    const results: Array<Readonly<IDinicEdge>> = []
    const { _edges, _edgesTot } = this
    for (let i = 0; i < _edgesTot; ++i) {
      const e = _edges[i]
      if (e.cap > 0 && e.flow === e.cap) results.push(e)
    }
    return results
  }

  protected _bfs(): boolean {
    const { _N, _source, _sink, _G, _edges, _Q, _dist } = this

    // Initialize the dist array.
    _dist.fill(-1, 0, _N)

    _Q.enqueue(_source)
    _dist[_source] = 0
    while (_Q.size > 0) {
      const o = _Q.dequeue()!
      for (const i of _G[o]) {
        const e = _edges[i]
        if (_dist[e.to] === -1 && e.cap > e.flow) {
          _dist[e.to] = _dist[o] + 1
          _Q.enqueue(e.to)
        }
      }
    }

    return _dist[_sink] !== -1
  }

  protected _dfs(o: number, mif: number): number {
    if (o === this._sink || mif === 0) return mif

    const { _G, _edges, _cur, _dist } = this

    let flow = 0
    for (let g = _G[o]; _cur[o] < g.length; ++_cur[o]) {
      const x = g[_cur[o]]
      const e = _edges[x]
      if (_dist[e.to] === _dist[o] + 1) {
        const remain = e.cap - e.flow
        const f = this._dfs(e.to, mif < remain ? mif : remain)
        if (f <= 0) continue

        e.flow += f
        _edges[x ^ 1].flow -= f
        flow += f
        // eslint-disable-next-line no-param-reassign
        mif -= f
        if (mif === 0) break
      }
    }
    return flow
  }
}
