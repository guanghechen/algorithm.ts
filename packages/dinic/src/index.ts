export * from './dinic'
export * from './types'
