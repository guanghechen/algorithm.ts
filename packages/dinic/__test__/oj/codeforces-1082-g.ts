import { Dinic } from '../../src'

export default solveCodeforces1082G

const dinic = new Dinic()
export function solveCodeforces1082G(
  nodes: number[],
  edges: Array<[u: number, v: number, weight: number]>,
): number {
  const n: number = nodes.length
  const m: number = edges.length

  const source = 0
  const target: number = n + m + 1
  dinic.init(source, target, n + m + 2)

  for (let i = 0; i < n; ++i) {
    const weight: number = nodes[i]
    dinic.addEdge(i + 1, target, weight)
  }

  let answer = 0
  for (let i = 0; i < m; ++i) {
    const [u, v, weight] = edges[i]
    const x = n + i
    answer += weight
    dinic.addEdge(source, x, weight)
    dinic.addEdge(x, u, Number.MAX_SAFE_INTEGER)
    dinic.addEdge(x, v, Number.MAX_SAFE_INTEGER)
  }
  answer -= dinic.maxflow()
  return answer
}
