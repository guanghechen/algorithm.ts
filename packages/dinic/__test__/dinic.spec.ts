import type { ITestData } from 'jest.setup'
import { loadCommonJsonFixtures } from 'jest.setup'
import { createDinic } from '../src'
import { solveCodeforces1082G } from './oj/codeforces-1082-g'
import { maxStudents } from './oj/maximum-students-taking-exam'

describe('basic', function () {
  test('simple', function () {
    const dinic = createDinic()
    dinic.init(0, 1, 4)
    dinic.addEdge(0, 2, 1)
    dinic.addEdge(0, 3, 2)
    dinic.addEdge(3, 1, 1)
    expect(dinic.maxFlow()).toEqual(1)
    dinic.solve(context => expect(context).toMatchSnapshot())
  })
})

describe('oj', function () {
  // https://codeforces.com/contest/1082/problem/G
  test('codeforces/1082/G', function () {
    type IParameters = Parameters<typeof solveCodeforces1082G>
    type IAnswer = ReturnType<typeof solveCodeforces1082G>
    const data: Array<ITestData<IParameters, IAnswer>> = loadCommonJsonFixtures('codeforces/1082/G')

    for (const { input, answer } of data) expect(solveCodeforces1082G(...input)).toEqual(answer)
  })

  // https://leetcode.com/problems/maximum-students-taking-exam/
  test('leetcode/maximum-students-taking-exam', function () {
    type IParameters = Parameters<typeof maxStudents>
    type IAnswer = ReturnType<typeof maxStudents>
    const data: Array<ITestData<IParameters, IAnswer>> = loadCommonJsonFixtures(
      'leetcode/maximum-students-taking-exam',
    )

    for (const { input, answer } of data) expect(maxStudents(...input)).toEqual(answer)
  })
})
