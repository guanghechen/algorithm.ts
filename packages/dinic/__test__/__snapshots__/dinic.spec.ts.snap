// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic simple 1`] = `
Object {
  "G": Array [
    Array [
      0,
      2,
    ],
    Array [
      5,
    ],
    Array [
      1,
    ],
    Array [
      3,
      4,
    ],
  ],
  "edgeTot": 6,
  "edges": Array [
    Object {
      "cap": 1,
      "flow": 0,
      "from": 0,
      "to": 2,
    },
    Object {
      "cap": 0,
      "flow": 0,
      "from": 2,
      "to": 0,
    },
    Object {
      "cap": 2,
      "flow": 1,
      "from": 0,
      "to": 3,
    },
    Object {
      "cap": 0,
      "flow": -1,
      "from": 3,
      "to": 0,
    },
    Object {
      "cap": 1,
      "flow": 1,
      "from": 3,
      "to": 1,
    },
    Object {
      "cap": 0,
      "flow": -1,
      "from": 1,
      "to": 3,
    },
  ],
}
`;
